<h1> Toronto data </h1>   
Scraping the Wikipedia page

In [1]:
import pandas as pd
df=pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")

Remove rows with borough "Not assigned"

In [2]:
toronto_data=df[0]
toronto_data=toronto_data[toronto_data['Borough']!='Not assigned'].sort_values(by=['Postcode'])
toronto_data

Postcode      Borough             Neighbourhood
11       M1B  Scarborough                     Rouge
12       M1B  Scarborough                   Malvern
29       M1C  Scarborough                Port Union
28       M1C  Scarborough                Rouge Hill
27       M1C  Scarborough            Highland Creek
42       M1E  Scarborough                 Guildwood
43       M1E  Scarborough               Morningside
44       M1E  Scarborough                 West Hill
53       M1G  Scarborough                    Woburn
62       M1H  Scarborough                 Cedarbrae
76       M1J  Scarborough       Scarborough Village
91       M1K  Scarborough      East Birchmount Park
92       M1K  Scarborough                   Ionview
93       M1K  Scarborough              Kennedy Park
108      M1L  Scarborough               Golden Mile
109      M1L  Scarborough                  Oakridge
107      M1L  Scarborough                  Clairlea
123      M1M  Scarborough                Cliffcrest
125      M1M  Scarborough  Scarborough Village West
124      M1M  Scarborough                 Cliffside
141      M1N  Scarborough            Cliffside West
140      M1N  Scarborough               Birch Cliff
153      M1P  Scarborough           Wexford Heights
151      M1P  Scarborough               Dorset Park
152      M1P  Scarborough   Scarborough Town Centre
164      M1R  Scarborough                  Maryvale
165      M1R  Scarborough                   Wexford
180      M1S  Scarborough                 Agincourt
192      M1T  Scarborough                  Sullivan
191      M1T  Scarborough            Clarks Corners
..       ...          ...                       ...
286      M8Z    Etobicoke            South of Bloor
285      M8Z    Etobicoke     Royal York South West
10       M9A    Etobicoke          Islington Avenue
22       M9B    Etobicoke                Cloverdale
26       M9B    Etobicoke           West Deane Park
25       M9B    Etobicoke          Princess Gardens
24       M9B    Etobicoke              Martin Grove
23       M9B    Etobicoke                 Islington
41       M9C    Etobicoke         Old Burnhamthorpe
40       M9C    Etobicoke             Markland Wood
39       M9C    Etobicoke                  Eringate
38       M9C    Etobicoke         Bloordale Gardens
122      M9L   North York             Humber Summit
139      M9M   North York                 Humberlea
138      M9M   North York                     Emery
150      M9N         York                    Weston
163      M9P    Etobicoke                 Westmount
178      M9R    Etobicoke          Richview Gardens
176      M9R    Etobicoke         Kingsview Village
179      M9R    Etobicoke              St. Phillips
177      M9R    Etobicoke      Martin Grove Gardens
228      M9V    Etobicoke            Albion Gardens
229      M9V    Etobicoke          Beaumond Heights
230      M9V    Etobicoke                Humbergate
231      M9V    Etobicoke                 Jamestown
232      M9V    Etobicoke               Mount Olive
234      M9V    Etobicoke             South Steeles
235      M9V    Etobicoke               Thistletown
233      M9V    Etobicoke               Silverstone
245      M9W    Etobicoke                 Northwest

[211 rows x 3 columns]

Create list of unique postcodes:

In [3]:
toronto_data
post_codes=toronto_data['Postcode'].drop_duplicates().to_list()
post_codes




['M1B',
 'M1C',
 'M1E',
 'M1G',
 'M1H',
 'M1J',
 'M1K',
 'M1L',
 'M1M',
 'M1N',
 'M1P',
 'M1R',
 'M1S',
 'M1T',
 'M1V',
 'M1W',
 'M1X',
 'M2H',
 'M2J',
 'M2K',
 'M2L',
 'M2M',
 'M2N',
 'M2P',
 'M2R',
 'M3A',
 'M3B',
 'M3C',
 'M3H',
 'M3J',
 'M3K',
 'M3L',
 'M3M',
 'M3N',
 'M4A',
 'M4B',
 'M4C',
 'M4E',
 'M4G',
 'M4H',
 'M4J',
 'M4K',
 'M4L',
 'M4M',
 'M4N',
 'M4P',
 'M4R',
 'M4S',
 'M4T',
 'M4V',
 'M4W',
 'M4X',
 'M4Y',
 'M5A',
 'M5B',
 'M5C',
 'M5E',
 'M5G',
 'M5H',
 'M5J',
 'M5K',
 'M5L',
 'M5M',
 'M5N',
 'M5P',
 'M5R',
 'M5S',
 'M5T',
 'M5V',
 'M5W',
 'M5X',
 'M6A',
 'M6B',
 'M6C',
 'M6E',
 'M6G',
 'M6H',
 'M6J',
 'M6K',
 'M6L',
 'M6M',
 'M6N',
 'M6P',
 'M6R',
 'M6S',
 'M7A',
 'M7R',
 'M7Y',
 'M8V',
 'M8W',
 'M8X',
 'M8Y',
 'M8Z',
 'M9A',
 'M9B',
 'M9C',
 'M9L',
 'M9M',
 'M9N',
 'M9P',
 'M9R',
 'M9V',
 'M9W']

For each value of post code, find part of dataframe with that post code. For each post code find unique values of neighbourhood and concate them into one string. If there is Not assigned value for neighbourhood, value of neighbourhood would be value of borough. Create dataframe with postcode, borough and joined values of neighbourhood. Append all dataframes in one final dataframe with right names of columns. 

In [4]:
data=pd.DataFrame()
for i in post_codes:
    df=toronto_data[toronto_data['Postcode']==i]
    nei=""
    br=0
    for j in df['Neighbourhood'].drop_duplicates():
        if j=='Not assigned':
            j=df['Borough'].to_list()[0]
        if br>0:
            nei+= ", "+j
            br+=1
        else:
            nei+=j
            br+=1

    tmp=pd.DataFrame()
    data=data.append(pd.DataFrame({'Postcode':[i],'Borough':[df['Borough'].to_list()[0]], 'Neighbourhood':[nei]}))


toronto_data=data.reset_index()[['Postcode','Borough','Neighbourhood']]

In [5]:
toronto_data.rename(columns={'Postcode':'PostalCode'}, inplace=True)
toronto_data

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
..         ...               ...   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                         Neighbourhood  
0                                       Rouge, Malvern  
1               Port Union, Rouge Hill, Highland Creek  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6          East Birchmount Park, Ionview, Kennedy Park  
7                      Golden Mile, Oakridge, Clairlea  
8      Cliffcrest, Scarborough Village West, Cliffside  
9                          Cliffside West, Birch Cliff  
10   Wexford Heights, Dorset Park, Scarborough Town...  
11                                   Maryvale, Wexford  
12                                           Agincourt  
13             Sullivan, Clarks Corners, Tam O'Shanter  
14   Milliken, Agincourt North, L'Amoreaux East, St...  
15                                     L'Amoreaux West  
16                                         Upper Rouge  
17                                   Hillcrest Village  
18                        Fairview, Oriole, Henry Farm  
19                                     Bayview Village  
20                            Silver Hills, York Mills  
21                             Willowdale, Newtonbrook  
22                                    Willowdale South  
23                                     York Mills West  
24                                     Willowdale West  
25                                           Parkwoods  
26                                     Don Mills North  
27                    Don Mills South, Flemingdon Park  
28     Wilson Heights, Downsview North, Bathurst Manor  
29                     Northwood Park, York University  
..  

Shape of final dataframe:

In [6]:
toronto_data.shape

(103, 3)

<h1> Geo data </h1>

This code is relevant for second Notebook:

Load geo data from csv file

In [7]:

df_geo=pd.read_csv("https://cocl.us/Geospatial_data")

Rename column to the same name as in toronto_data dataframe

In [8]:
df_geo.rename(columns={'Postal Code': 'PostalCode'}, inplace=True)
df_geo


PostalCode   Latitude  Longitude
0          M1B  43.806686 -79.194353
1          M1C  43.784535 -79.160497
2          M1E  43.763573 -79.188711
3          M1G  43.770992 -79.216917
4          M1H  43.773136 -79.239476
5          M1J  43.744734 -79.239476
6          M1K  43.727929 -79.262029
7          M1L  43.711112 -79.284577
8          M1M  43.716316 -79.239476
9          M1N  43.692657 -79.264848
10         M1P  43.757410 -79.273304
11         M1R  43.750072 -79.295849
12         M1S  43.794200 -79.262029
13         M1T  43.781638 -79.304302
14         M1V  43.815252 -79.284577
15         M1W  43.799525 -79.318389
16         M1X  43.836125 -79.205636
17         M2H  43.803762 -79.363452
18         M2J  43.778517 -79.346556
19         M2K  43.786947 -79.385975
20         M2L  43.757490 -79.374714
21         M2M  43.789053 -79.408493
22         M2N  43.770120 -79.408493
23         M2P  43.752758 -79.400049
24         M2R  43.782736 -79.442259
25         M3A  43.753259 -79.329656
26         M3B  43.745906 -79.352188
27         M3C  43.725900 -79.340923
28         M3H  43.754328 -79.442259
29         M3J  43.767980 -79.487262
..         ...        ...        ...
73         M6C  43.693781 -79.428191
74         M6E  43.689026 -79.453512
75         M6G  43.669542 -79.422564
76         M6H  43.669005 -79.442259
77         M6J  43.647927 -79.419750
78         M6K  43.636847 -79.428191
79         M6L  43.713756 -79.490074
80         M6M  43.691116 -79.476013
81         M6N  43.673185 -79.487262
82         M6P  43.661608 -79.464763
83         M6R  43.648960 -79.456325
84         M6S  43.651571 -79.484450
85         M7A  43.662301 -79.389494
86         M7R  43.636966 -79.615819
87         M7Y  43.662744 -79.321558
88         M8V  43.605647 -79.501321
89         M8W  43.602414 -79.543484
90         M8X  43.653654 -79.506944
91         M8Y  43.636258 -79.498509
92         M8Z  43.628841 -79.520999
93         M9A  43.667856 -79.532242
94         M9B  43.650943 -79.554724
95         M9C  43.643515 -79.577201
96         M9L  43.756303 -79.565963
97         M9M  43.724766 -79.532242
98         M9N  43.706876 -79.518188
99         M9P  43.696319 -79.532242
100        M9R  43.688905 -79.554724
101        M9V  43.739416 -79.588437
102        M9W  43.706748 -79.594054

[103 rows x 3 columns]

Merge geo_data and toronto_data dataframes

In [9]:
toronto_geo_data=pd.merge(toronto_data,df_geo, on=['PostalCode'])

In [10]:
toronto_geo_data

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
..         ...               ...   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                       Rouge, Malvern  43.806686 -79.194353  
1               Port Union, Rouge Hill, Highland Creek  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
5                                  Scarborough Village  43.744734 -79.239476  
6          East Birchmount Park, Ionview, Kennedy Park  43.727929 -79.262029  
7                      Golden Mile, Oakridge, Clairlea  43.711112 -79.284577  
8      Cliffcrest, Scarborough Village West, Cliffside  43.716316 -79.239476  
9                          Cliffside West, Birch Cliff  43.692657 -79.264848  
10   Wexford Heights, Dorset Park, Scarborough Town...  43.757410 -79.273304  
11                                   Maryvale, Wexford  43.750072 -79.295849  
12                                           Agincourt  43.794200 -79.262029  
13             Sullivan, Clarks Corners, Tam O'Shanter  43.781638 -79.304302  
14   Milliken, Agincourt North, L'Amoreaux East, St...  43.815252 -79.284577  
15                                     L'Amoreaux West  43.799525 -79.318389  
16                                         Upper Rouge  43.836125 -79.205636  
17                                   Hillcrest Village  43.803762 -79.363452  
18                        Fairview, Oriole, Henry Farm  43.778517 -79.346556  
19                                     Bayview Village  43.786947 -79.385975  
20                            Silver Hills, York Mills  43.757490 -79.374714  
21                             Wi

<h1> Neighborhoods New York </h1>
All below is for last notebook in segmentation and clustering project

<h2> Toronto map </h2>

In [14]:
import numpy as np # library to handle data in a vectorized manner

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library
!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
neighborhoods=toronto_geo_data

Solving environment: done

# All requested packages already installed.

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.49         |             py_0          32 KB  conda-forge
    geopy-1.20.0               |             py_0          57 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          90 KB

The following NEW packages will be INSTALLED:

    geographiclib: 1.49-py_0   conda-forge
    geopy:         1.20.0-py_0 conda-forge


geographiclib-1.49   | 32 KB     | ##################################### | 100% 
geopy-1.20.0         | 57 KB     | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executi

<b> Finding boroughs with Toronto as part of name </b>

In [23]:
neighborhoods=neighborhoods[neighborhoods['Borough'].str.contains("Toronto")]
neighborhoods

PostalCode           Borough  \
37        M4E      East Toronto   
41        M4K      East Toronto   
42        M4L      East Toronto   
43        M4M      East Toronto   
44        M4N   Central Toronto   
45        M4P   Central Toronto   
46        M4R   Central Toronto   
47        M4S   Central Toronto   
48        M4T   Central Toronto   
49        M4V   Central Toronto   
50        M4W  Downtown Toronto   
51        M4X  Downtown Toronto   
52        M4Y  Downtown Toronto   
53        M5A  Downtown Toronto   
54        M5B  Downtown Toronto   
55        M5C  Downtown Toronto   
56        M5E  Downtown Toronto   
57        M5G  Downtown Toronto   
58        M5H  Downtown Toronto   
59        M5J  Downtown Toronto   
60        M5K  Downtown Toronto   
61        M5L  Downtown Toronto   
63        M5N   Central Toronto   
64        M5P   Central Toronto   
65        M5R   Central Toronto   
66        M5S  Downtown Toronto   
67        M5T  Downtown Toronto   
68        M5V  Downtown Toronto   
69        M5W  Downtown Toronto   
70        M5X  Downtown Toronto   
75        M6G  Downtown Toronto   
76        M6H      West Toronto   
77        M6J      West Toronto   
78        M6K      West Toronto   
82        M6P      West Toronto   
83        M6R      West Toronto   
84        M6S      West Toronto   
87        M7Y      East Toronto   

                                        Neighbourhood   Latitude  Longitude  
37                                        The Beaches  43.676357 -79.293031  
41                       The Danforth West, Riverdale  43.679557 -79.352188  
42                     The Beaches West, India Bazaar  43.668999 -79.315572  
43                                    Studio District  43.659526 -79.340923  
44                                      Lawrence Park  43.728020 -79.388790  
45                                   Davisville North  43.712751 -79.390197  
46                                 North Toronto West  43.715383 -79.405678  
47                                         Davisville  43.704324 -79.388790  
48                        Moore Park, Summerhill East  43.689574 -79.383160  
49  South Hill, Rathnelly, Summerhill West, Deer P...  43.686412 -79.400049  
50                                           Rosedale  43.679563 -79.377529  
51                        Cabbagetown, St. James Town  43.667967 -79.367675  
52                               Church and Wellesley  43.665860 -79.383160  
53                          Regent Park, Harbourfront  43.654260 -79.360636  
54                           Ryerson, Garden District  43.657162 -79.378937  
55                                     St. James Town  43.651494 -79.375418  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
58                           King, Adelaide, Richmond  43.650571 -79.384568  
59  Toronto Islands, Harbourfront East, Union Station  43.640816 -79.381752  
60           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
61                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
63                                           Roselawn  43.711695 -79.416936  
64                Forest Hill North, Forest Hill West  43.696948 -79.411307  
65                The Annex, Yorkville, North Midtown  43.672710 -79.405678  
66                     University of Toronto, Harbord  43.662696 -79.400049  
67          Kensington Market, Grange Park, Chinatown  43.653206 -79.400049  
68  Railway Lands, South Niagara, Harbourfront Wes...  43.628947 -79.394420  
69                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846  
70             Underground city, First Canadian Place  43.648429 -79.382280  
75                                           Christie  43.669542 -79.422564  
76                       Dovercourt Village, Dufferin  43.669005 -79.442259  
77                           Trinity, Little Portugal  43.64792

Finding of latitude and longitude of Toronto

In [24]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [17]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)


# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

<h2>  Foursquare API exploring </h2>

Creating Foursquare credentials

In [27]:
CLIENT_ID = 'LKHH1IZTYFREOCCLNGRUV0XSIGNYGHRF2SXASHVGAY2L1T1L' 
CLIENT_SECRET = 'FJFCDDKCH0SGMYV2XGIACGULJHYF2YPN2JXNX3KLUKVOFSWA'
VERSION = '20180605' # Foursquare API version

Reseting index of dataframe

In [31]:
neighborhoods.reset_index(inplace=True)

<h3> Exploring first neighborhood in dataframe </h3>

Finding of location data and name 

In [40]:
neighborhood_latitude = neighborhoods.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = neighborhoods.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = neighborhoods.loc[0, 'Neighbourhood'] # neighborhood name

Getting foursquare data for max 100 venuse in radius of 500 meters

In [41]:
# type your answer here

LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d5150917ca06c0030e41284'},
 'response': {'headerLocation': 'The Beaches',
  'headerFullLocation': 'The Beaches, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 5,
  'suggestedBounds': {'ne': {'lat': 43.680857404499996,
    'lng': -79.28682091449052},
   'sw': {'lat': 43.67185739549999, 'lng': -79.29924148550948}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bd461bc77b29c74a07d9282',
       'name': 'Glen Manor Ravine',
       'location': {'address': 'Glen Manor',
        'crossStreet': 'Queen St.',
        'lat': 43.67682094413784,
        'lng': -79.29394208780985,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.67682094413784,
          'lng': -79.29394208780985}],
        'distanc

In [42]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Parsing of json results

In [43]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat  \
0                   Glen Manor Ravine                 Trail  43.676821   
1  The Big Carrot Natural Food Market     Health Food Store  43.678879   
2                 Grover Pub and Grub                   Pub  43.679181   
3                 Glen Stewart Ravine  Other Great Outdoors  43.676300   
4                       Upper Beaches          Neighborhood  43.680563   

         lng  
0 -79.293942  
1 -79.297734  
2 -79.297215  
3 -79.294784  
4 -79.292869

In [44]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

5 venues were returned by Foursquare.


<h3> Creating function for all neighbors </h3>


In [46]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [47]:
toronto_venues = getNearbyVenues(names=neighborhoods['Neighbourhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
South Hill, Rathnelly, Summerhill West, Deer Park, Forest Hill SE
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Regent Park, Harbourfront
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
King, Adelaide, Richmond
Toronto Islands, Harbourfront East, Union Station
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
The Annex, Yorkville, North Midtown
University of Toronto, Harbord
Kensington Market, Grange Park, Chinatown
Railway Lands, South Niagara, Harbourfront West, Island airport, Bathurst Quay, CN Tower, King and Spadina
Stn A PO Boxes 25 The Esplanade
Underground city, First Canadian Place
Christie
Dovercourt Village, Dufferin
Trinity, Little Portugal
Exhibition Place, Brockton, Parkdale Village
High Park, The 

<h4> Resulting dataframe has size: </h4>

In [51]:
toronto_venues.shape

(1690, 7)

Dataframe has 1690 rows. Let's see how much venues has each neighbour:

In [54]:
res=toronto_venues.groupby('Neighborhood').count()
res

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            56   
Business Reply Mail Processing Centre 969 Eastern                      16   
Cabbagetown, St. James Town                                            46   
Central Bay Street                                                     82   
Christie                                                               16   
Church and Wellesley                                                   82   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                       10   
Dovercourt Village, Dufferin                                           15   
Exhibition Place, Brockton, Parkdale Village                           23   
Forest Hill North, Forest Hill West                                     4   
High Park, The Junction South                                          23   
Kensington Market, Grange Park, Chinatown                             100   
King, Adelaide, Richmond                                              100   
Lawrence Park                                                           4   
Moore Park, Summerhill East                                             3   
North Toronto West                                                     20   
Railway Lands, South Niagara, Harbourfront West...                     16   
Regent Park, Harbourfront                                              45   
Roncesvalles, Parkdale                                                 15   
Rosedale                                                                5   
Roselawn                                                                3   
Runnymede, Swansea                                                     38   
Ryerson, Garden District                                              100   
South Hill, Rathnelly, Summerhill West, Deer Pa...                     15   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        96   
Studio District                                                        40   
The Annex, Yorkville, North Midtown                                    23   
The Beaches                                                             5   
The Beaches West, India Bazaar                                         17   
The Danforth West, Riverdale                                           41   
Toronto Dominion Centre, Design Exchange                              100   
Toronto Islands, Harbourfront East, Union Station                     100   
Trinity, Little Portugal                                               64   
Underground city, First Canadian Place                                100   
University of Toronto, Harbord                                         33   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             56   
Business Reply Mail Processing Centre 969 Eastern                       16   
Cabbagetown, St. James Town                                             46   
Central Bay Street                                                      82   
Christie                                                                16   
Church and Wellesley                                                    82   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              34   
Davisville North                                                        10   
Dovercourt Village, Dufferin                                            15   
Exhibition Place, Brockton, Parkdal

In [56]:
res[res.Venue==res.Venue.max()]

Neighborhood Latitude  \
Neighborhood                                                               
Commerce Court, Victoria Hotel                                       100   
Kensington Market, Grange Park, Chinatown                            100   
King, Adelaide, Richmond                                             100   
Ryerson, Garden District                                             100   
St. James Town                                                       100   
Toronto Dominion Centre, Design Exchange                             100   
Toronto Islands, Harbourfront East, Union Station                    100   
Underground city, First Canadian Place                               100   

                                                   Neighborhood Longitude  \
Neighborhood                                                                
Commerce Court, Victoria Hotel                                        100   
Kensington Market, Grange Park, Chinatown                             100   
King, Adelaide, Richmond                                              100   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Toronto Dominion Centre, Design Exchange                              100   
Toronto Islands, Harbourfront East, Union Station                     100   
Underground city, First Canadian Place                                100   

                                                   Venue  Venue Latitude  \
Neighborhood                                                               
Commerce Court, Victoria Hotel                       100             100   
Kensington Market, Grange Park, Chinatown            100             100   
King, Adelaide, Richmond                             100             100   
Ryerson, Garden District                             100             100   
St. James Town                                       100             100   
Toronto Dominion Centre, Design Exchange             100             100   
Toronto Islands, Harbourfront East, Union Station    100             100   
Underground city, First Canadian Place               100             100   

                                                   Venue Longitude  \
Neighborhood                                                         
Commerce Court, Victoria Hotel                                 100   
Kensington Market, Grange Park, Chinatown                      100   
King, Adelaide, Richmond                                       100   
Ryerson, Garden District                                       100   
St. James Town                                                 100   
Toronto Dominion Centre, Design Exchange                       100   
Toronto Islands, Harbourfront East, Union Station              100   
Underground city, First Canadian Place                         100   

                                                   Venue Category  
Neighborhood                                                       
Commerce Court, Victoria Hotel                                100  
Kensington Market, Grange Park, Chinatown                     100  
King, Adelaide, Richmond                                      100  
Ryerson, Garden District                                      100  
St. James Town                                                100  
Toronto Dominion Centre, Design Exchange                      100  
Toronto Islands, Harbourfront East, Union Station             100  
Underground city, First Canadian Place                        100

There are few Nesighborhoods with max defined number of venues

In [57]:
res[res.Venue==res.Venue.min()]

Neighborhood Latitude  Neighborhood Longitude  \
Neighborhood                                                                 
Moore Park, Summerhill East                      3                       3   
Roselawn                                         3                       3   

                             Venue  Venue Latitude  Venue Longitude  \
Neighborhood                                                          
Moore Park, Summerhill East      3               3                3   
Roselawn                         3               3                3   

                             Venue Category  
Neighborhood                                 
Moore Park, Summerhill East               3  
Roselawn                                  3

The minimum number of venues is 3

<h2> Analysis of each Neighborhood </h2>


In [61]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0             0         0            0                    0                 0   
1             0         0            0                    0                 0   
2             0         0            0                    0                 0   
3             0         0            0                    0                 0   
4             0         0            0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       0     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Bistro  Boat or Ferry  Bookstore  \
0                0         0           0       0              0          0   
1                0         0           0       0              0          0   
2                0         0           0       0              0          0   
3                0         0           0       0              0          0   
4                0         0           0       0              0          0   

   Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0         0                     0               0        0                0   
1         0                     0               0        0                0   
2         0                     0               0        0                0   
3         0                     0               0        0                0   
4         0                     0               0        0                0   

   Building  Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0         0             0              0         0        0     0   
1         0             0              0         0        0     0   
2         0             0              0         0        0     0   
3         0             0              0         0        0     0   
4         0             0              0         0        0     0   

   Cajun / Creole Restaurant  Camera Store  Caribbean Restaurant  Cheese Shop  \
0                          0             0                     0            0   
1                          0             0                     0            0   
2                          0             0                     0            0   
3                      

<b> Grouping data by neighborhood </b>

In [78]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1   Business Reply Mail Processing Centre 969 Eastern     0.000000   
2                         Cabbagetown, St. James Town     0.000000   
3                                  Central Bay Street     0.012195   
4                                            Christie     0.000000   
5                                Church and Wellesley     0.012195   
6                      Commerce Court, Victoria Hotel     0.000000   
7                                          Davisville     0.000000   
8                                    Davisville North     0.000000   
9                        Dovercourt Village, Dufferin     0.000000   
10       Exhibition Place, Brockton, Parkdale Village     0.043478   
11                Forest Hill North, Forest Hill West     0.000000   
12                      High Park, The Junction South     0.000000   
13          Kensington Market, Grange Park, Chinatown     0.000000   
14                           King, Adelaide, Richmond     0.000000   
15                                      Lawrence Park     0.000000   
16                        Moore Park, Summerhill East     0.000000   
17                                 North Toronto West     0.050000   
18  Railway Lands, South Niagara, Harbourfront Wes...     0.000000   
19                          Regent Park, Harbourfront     0.000000   
20                             Roncesvalles, Parkdale     0.000000   
21                                           Rosedale     0.000000   
22                                           Roselawn     0.000000   
23                                 Runnymede, Swansea     0.000000   
24                           Ryerson, Garden District     0.000000   
25  South Hill, Rathnelly, Summerhill West, Deer P...     0.000000   
26                                     St. James Town     0.000000   
27                    Stn A PO Boxes 25 The Esplanade     0.000000   
28                                    Studio District     0.025000   
29                The Annex, Yorkville, North Midtown     0.000000   
30                                        The Beaches     0.000000   
31                     The Beaches West, India Bazaar     0.000000   
32                       The Danforth West, Riverdale     0.024390   
33           Toronto Dominion Centre, Design Exchange     0.000000   
34  Toronto Islands, Harbourfront East, Union Station     0.000000   
35                           Trinity, Little Portugal     0.015625   
36             Underground city, First Canadian Place     0.000000   
37                     University of Toronto, Harbord     0.000000   

    Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0.000000   0.0000              0.0000        0.0000   
1            0.000000   0.0000              0.0000        0.0000   
2            0.000000   0.0000              0.0000        0.0000   
3            0.000000   0.0000              0.0000        0.0000   
4            0.000000   0.0000              0.0000        0.0000   
5            0.012195   0.0000              0.0000        0.0000   
6            0.000000   0.0000              0.0000        0.0000   
7            0.000000   0.0000              0.0000        0.0000   
8            0.000000   0.0000              0.0000        0.0000   
9            0.000000   0.0000              0.0000        0.0000   
10           0.000000   0.0000              0.0000        0.0000   
11           0.000000   0.0000              0.0000        0.0000   
12           0.000000   0.0000              0.0000        0.0000   
13           0.000000   0.0000              0.0000        0.0000   
14           0.000000   0.0000              0.0000        0.0000   
15           0.000000   0.0000              0.0000        0.0000   
16           0.000000   0.0000              0.0000        0.0000   
17           0.000000   0.0000              0.0000        0.0000   
18           0.000

In [63]:
#shape of grouped data
toronto_grouped.shape

(38, 232)

<h1>Clustering neighborhoods</h1>

In [94]:
# set number of clusters
kclusters = 4

toronto_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:100] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 2, 3, 0, 0, 0, 0, 3,
       1, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [97]:
# add clustering labels
toronto_grouped.insert(0, 'Cluster Labels', kmeans.labels_)


In [98]:

neighborhoods.rename(columns={'Neighbourhood':'Neighborhood'}, inplace=True)

In [99]:

toronto_merged = neighborhoods

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(toronto_grouped.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() 

index PostalCode          Borough                    Neighborhood  \
0     37        M4E     East Toronto                     The Beaches   
1     41        M4K     East Toronto    The Danforth West, Riverdale   
2     42        M4L     East Toronto  The Beaches West, India Bazaar   
3     43        M4M     East Toronto                 Studio District   
4     44        M4N  Central Toronto                   Lawrence Park   

    Latitude  Longitude  Cluster Labels  Yoga Studio  Afghan Restaurant  \
0  43.676357 -79.293031               2      0.00000                0.0   
1  43.679557 -79.352188               0      0.02439                0.0   
2  43.668999 -79.315572               0      0.00000                0.0   
3  43.659526 -79.340923               0      0.02500                0.0   
4  43.728020 -79.388790               2      0.00000                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0               0.0              0.00000           0.0       0.0          0.0   
1               0.0              0.02439           0.0       0.0          0.0   
2               0.0              0.00000           0.0       0.0          0.0   
3               0.0              0.05000           0.0       0.0          0.0   
4               0.0              0.00000           0.0       0.0          0.0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                  0.0               0.0                 0.0            0.0   
1                  0.0               0.0                 0.0            0.0   
2                  0.0               0.0                 0.0            0.0   
3                  0.0               0.0                 0.0            0.0   
4                  0.0               0.0                 0.0            0.0   

   BBQ Joint  Baby Store  Bagel Shop   Bakery   Bank    Bar  Baseball Stadium  \
0        0.0         0.0         0.0  0.00000  0.000  0.000               0.0   
1        0.0         0.0         0.0  0.02439  0.000  0.000               0.0   
2        0.0         0.0         0.0  0.00000  0.000  0.000               0.0   
3        0.0         0.0         0.0  0.05000  0.025  0.025               0.0   
4        0.0         0.0         0.0  0.00000  0.000  0.000               0.0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  Bistro  \
0                 0.0    0.0              0.0       0.0         0.0     0.0   
1                 0.0    0.0              0.0       0.0         0.0     0.0   
2                 0.0    0.0              0.0       0.0         0.0     0.0   
3                 0.0    0.0              0.0       0.0         0.0     0.0   
4                 0.0    0.0              0.0       0.0         0.0     0.0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0            0.0    0.00000       0.0                   0.0             0.0   
1            0.0    0.02439       0.0                   0.0             0.0   
2            0.0    0.00000       0.0                   0.0             0.0   
3            0.0    0.02500       0.0                   0.0             0.0   
4            0.0    0.00000       0.0                   0.0             0.0   

    Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0  0.000000          0.00000       0.0      0.000000       0.000000      0.00   
1  0.024390          0.02439       0.0      0.000000       0.000000    

In [100]:
toronto_merged.shape

(38, 238)

<h2> Visualisation of clusters </h2>

In [101]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters